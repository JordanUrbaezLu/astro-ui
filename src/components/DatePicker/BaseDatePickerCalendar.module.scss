@use "../../utility/globals";

.datePickerCalendarContainer {
  background: globals.$primary-background-color;
  border-radius: 10px;
  box-shadow: rgb(0 0 0 / 35%) 0 5px 15px;
  color: globals.$primary-text-color;
  display: flex;
  flex-direction: column;
  font-family: globals.$font-family;
  margin-top: 2px;
  position: absolute;
  width: 360px;
}

.datePickerCalendarNavigationContainer {
  display: flex;
  justify-content: space-between;
  padding: 8px;
}

.datePickerCalendarNavigationMonth {
  align-items: center;
  display: flex;
  user-select: none;
}

.datePickerCalendarHeaderContainer {
  display: flex;
  padding: 2px 12px;
}

.datePickerCalendarHeaderLabel {
  flex: 1 1 0px;
  text-align: center;
  user-select: none;
}

.datePickerCalendarBodyContainer {
  display: flex;
  flex-wrap: wrap;
  padding: 0 12px;
}

.datePickerCalendarDayContainer {
  flex: 1 0 13%;
}

.datePickerCalendarDay {
  align-items: center;
  background-color: transparent;
  border: 2px solid transparent;
  border-radius: 1000px;
  color: globals.$primary-text-color;
  cursor: pointer;
  display: flex;
  height: 34px;
  justify-content: center;
  margin: auto;
  outline: none;
  user-select: none;
  width: 34px;

  &.selected {
    background-color: globals.$primary-border-color;
    border-color: globals.$primary-border-color !important;
  }

  &.today {
    border: 2px solid globals.$primary-text-color !important;
  }

  &:hover,
  &:focus {
    background-color: globals.$primary-background-color-hover-and-focus;
    border: 2px solid
      globals.$primary-background-color-hover-and-focus;
  }

  &:active {
    background-color: globals.$primary-background-color-active;
    border: 2px solid globals.$primary-background-color-active;
  }
}

.buttonContainer {
  align-items: center;
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  padding: 6px 4px;
  width: 100%;
}
