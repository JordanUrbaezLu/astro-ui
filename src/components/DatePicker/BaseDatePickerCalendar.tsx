/* istanbul ignore file */
import * as React from "react";
import styles from "./BaseDatePickerCalendar.module.scss";
import classNames from "classnames";
import { useOnClickOutside, useOnKeyDown } from "../../hooks";
import {
  TfiAngleLeft,
  TfiAngleRight,
  TfiAngleDoubleRight,
  TfiAngleDoubleLeft,
} from "react-icons/tfi";
import {
  getCalendarDays,
  getNextDay,
  getNextWeek,
  getPreviousDay,
  getPreviousWeek,
} from "../../utility/getDatePickerCalendarUtilities";
import { DatePickerContext } from "./DatePickerContext";
import { Button } from "../Button";
import { FocusTrap } from "../FocusTrap/FocusTrap";
import { IconButton } from "../IconButton/IconButton";

export interface BaseDatePickerCalendarProps {
  /**
   * The callback fired when the DatePickerCalendar requests to close
   */
  onClose: () => void;
  /**
   * The trigger ref for the DatePicker
   */
  triggerRef: React.RefObject<HTMLButtonElement>;
}

/**
 * @private
 */
export const BaseDatePickerCalendar: React.FC<BaseDatePickerCalendarProps> =
  (props) => {
    const { onClose, triggerRef } = props;

    const { selectedDate, setSelectedDate } =
      React.useContext(DatePickerContext);
    const [date, setDate] = React.useState<Date>(
      selectedDate ?? new Date()
    );
    const [focusedDate, setFocusedDate] = React.useState<number>(0);
    const [days, setDays] = React.useState<Array<number | null>>([]);
    const daysRefs = React.useRef<Array<HTMLButtonElement | null>>(
      []
    );
    const calendarRef = React.useRef<HTMLDivElement>(null);

    const headerLabels = [
      "Sun",
      "Mon",
      "Tue",
      "Wed",
      "Thu",
      "Fri",
      "Sat",
    ];

    const increaseMonth = () => {
      setDate(
        new Date(
          date.getFullYear(),
          date.getMonth() + 1,
          date.getDate()
        )
      );
    };

    const increaseYear = () => {
      setDate(
        new Date(
          date.getFullYear() + 1,
          date.getMonth(),
          date.getDate()
        )
      );
    };

    const decreaseMonth = () => {
      setDate(
        new Date(
          date.getFullYear(),
          date.getMonth() - 1,
          date.getDate()
        )
      );
    };

    const decreaseYear = () => {
      setDate(
        new Date(
          date.getFullYear() - 1,
          date.getMonth(),
          date.getDate()
        )
      );
    };

    React.useEffect(() => {
      setDays(getCalendarDays(date));
      setFocusedDate(date.getDate());
    }, [date]);

    React.useEffect(() => {
      daysRefs.current[focusedDate]?.focus();
    }, [focusedDate]);

    useOnClickOutside([calendarRef, triggerRef], onClose);

    useOnKeyDown(["Escape"], onClose);

    return (
      <div
        className={styles.datePickerCalendarContainer}
        ref={calendarRef}
      >
        <FocusTrap>
          <div
            className={styles.datePickerCalendarNavigationContainer}
          >
            <div style={{ display: "flex", gap: "2px" }}>
              <IconButton
                content={<TfiAngleDoubleLeft />}
                size="medium"
                onClick={decreaseYear}
              />
              <IconButton
                content={<TfiAngleLeft />}
                size="medium"
                onClick={decreaseMonth}
              />
            </div>
            <div className={styles.datePickerCalendarNavigationMonth}>
              {`${date.toLocaleString("en-US", {
                month: "long",
              })} ${date.getFullYear()}`}
            </div>
            <div style={{ display: "flex", gap: "2px" }}>
              <IconButton
                content={<TfiAngleRight />}
                size="medium"
                onClick={increaseMonth}
              />
              <IconButton
                content={<TfiAngleDoubleRight />}
                size="medium"
                onClick={increaseYear}
              />
            </div>
          </div>
          <div className={styles.datePickerCalendarHeaderContainer}>
            {headerLabels.map((label, index) => {
              return (
                <div
                  className={styles.datePickerCalendarHeaderLabel}
                  key={index}
                >
                  {label}
                </div>
              );
            })}
          </div>
          <div className={styles.datePickerCalendarBodyContainer}>
            {days.map((day, index) => {
              if (day !== null) {
                const isSelected =
                  new Date(
                    date.getFullYear(),
                    date.getMonth(),
                    day
                  ).getTime() === selectedDate?.getTime();

                const isToday =
                  new Date(
                    date.getFullYear(),
                    date.getMonth(),
                    day
                  ).getTime() ===
                  new Date(
                    new Date().getFullYear(),
                    new Date().getMonth(),
                    new Date().getDate()
                  ).getTime();

                return (
                  <div
                    className={styles.datePickerCalendarDayContainer}
                    key={index}
                  >
                    <button
                      className={classNames(
                        styles.datePickerCalendarDay,
                        isSelected && styles.selected,
                        isToday && styles.today
                      )}
                      tabIndex={day === date.getDate() ? 0 : -1}
                      onClick={() => {
                        setSelectedDate(
                          new Date(
                            date.getFullYear(),
                            date.getMonth(),
                            day
                          )
                        );
                        onClose();
                      }}
                      onKeyDown={(event) => {
                        const { code } = event;

                        const keyCodesToGet = {
                          ArrowUp: getPreviousWeek,
                          ArrowDown: getNextWeek,
                          ArrowLeft: getPreviousDay,
                          ArrowRight: getNextDay,
                        };

                        if (code in keyCodesToGet) {
                          event.preventDefault();
                          setDate(keyCodesToGet[code]);
                        }

                        if (code === "Enter") {
                          setSelectedDate(
                            new Date(
                              date.getFullYear(),
                              date.getMonth(),
                              day
                            )
                          );
                          onClose();
                        }
                      }}
                      ref={(el) => (daysRefs.current[day] = el)}
                    >
                      {day}
                    </button>
                  </div>
                );
              } else {
                return (
                  <div
                    className={styles.datePickerCalendarDayContainer}
                    key={index}
                  />
                );
              }
            })}
          </div>
          <div className={styles.buttonContainer}>
            <Button onClick={() => setSelectedDate(undefined)}>
              Clear
            </Button>
          </div>
        </FocusTrap>
      </div>
    );
  };

BaseDatePickerCalendar.displayName = "BaseDatePickerCalendar";
