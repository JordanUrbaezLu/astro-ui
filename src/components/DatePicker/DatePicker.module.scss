@use "../../utility/globals";

.customInput {
  width: 165px;
}

.datePickerIconContainer {
  border: 3px solid transparent;
  border-radius: 999px;
  color: globals.$primary-background-color;
  cursor: pointer;
  outline: none;
  padding: 4px;

  &:hover,
  &:focus {
    border: 3px solid
      globals.$primary-background-color-hover-and-focus;
  }

  &:active {
    border: 3px solid globals.$primary-background-color-active;
    color: globals.$primary-background-color-active;
  }
}
