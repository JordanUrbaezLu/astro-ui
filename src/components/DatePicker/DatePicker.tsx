/* istanbul ignore file */
import * as React from "react";
import { AiFillCalendar } from "react-icons/ai";
import { getDateString } from "../../utility/getDatePickerCalendarUtilities";
import { DatePickerCalendar } from "./DatePickerCalendar";
import { TextField } from "../TextField/TextField";
import styles from "./DatePicker.module.scss";
import { DatePickerContext } from "./DatePickerContext";
import { IconButton } from "../IconButton/IconButton";

export interface DatePickerProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The initial selected date for the DatePicker
   */
  selected?: Date;
  /**
   * The callback fired when requested to change the value for the DatePicker
   */
  setValue: React.Dispatch<React.SetStateAction<string>>;
  /**
   * The value for the DatePicker
   */
  value?: string;
}

/**
 * @public
 *
 * @description
 *
 * The Date Picker allows users to select a date on a calendar.
 */
export const DatePicker: React.FC<DatePickerProps> = (props) => {
  const { className, selected, setValue, value, ...rest } = props;

  const triggerRef = React.useRef<HTMLButtonElement>(null);

  const [selectedDate, setSelectedDate] =
    React.useState<Date | undefined>(selected);
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  React.useEffect(() => {
    setValue(getDateString(selectedDate));
  }, [selectedDate]);

  return (
    <DatePickerContext.Provider
      value={{
        selectedDate: selectedDate,
        setSelectedDate: setSelectedDate,
      }}
    >
      <div className={className} {...rest}>
        <TextField
          className={styles.customInput}
          onChange={() => undefined}
          value={value}
          label="Select a Date"
          trailingIcon={
            <IconButton
              content={<AiFillCalendar />}
              size="medium"
              onClick={() => setIsOpen(!isOpen)}
              ref={triggerRef}
            />
          }
        />
        <DatePickerCalendar
          isOpen={isOpen}
          onClose={() => setIsOpen(false)}
          triggerRef={triggerRef}
        />
      </div>
    </DatePickerContext.Provider>
  );
};

DatePicker.displayName = "DatePicker";
