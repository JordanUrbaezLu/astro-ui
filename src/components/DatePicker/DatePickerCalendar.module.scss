.enter {
  opacity: 0;
  transform: translateY(-6px);
}

.enterActive {
  opacity: 1;
  transform: translateY(0);
  transition: all 200ms ease-in-out;
}

.exit {
  opacity: 1;
  transform: translateY(0);
}

.exitActive {
  opacity: 0;
  transform: translateY(-6px);
  transition: all 200ms ease-in-out;
}
