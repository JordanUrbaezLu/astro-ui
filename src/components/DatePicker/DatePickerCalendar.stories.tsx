import { Story } from "@storybook/react";
import {
  DatePickerCalendar,
  DatePickerCalendarProps,
} from "./DatePickerCalendar";

export default {
  component: DatePickerCalendar,
  title: "Components/DatePickerCalendar",
};

const Template: Story<
  Omit<DatePickerCalendarProps, "isOpen" | "onSetDate">
> = (args) => {
  return <DatePickerCalendar isOpen onSetDate={() => null} {...args} />;
};

export const Default = Template.bind({});
Default.args = {
  selectedDate: new Date(2019, 5, 9),
};
Default.parameters = {
  layout: "padded",
};
