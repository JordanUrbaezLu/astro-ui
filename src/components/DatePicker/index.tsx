/* istanbul ignore file */
export * from "./DatePicker";
