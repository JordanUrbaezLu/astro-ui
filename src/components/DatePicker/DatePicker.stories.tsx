import { Story } from "@storybook/react";
import * as React from "react";
import { DatePicker, DatePickerProps } from "./DatePicker";
import { within, userEvent } from "@storybook/testing-library";
import { allowLayoutCalculations } from "../../../.storybook/utilities/allowLayoutCalculations";

export default {
  component: DatePicker,
  parameters: {
    chromatic: { delay: 2000 },
  },
  title: "Components/Inputs/DatePicker",
};

const Template: Story<DatePickerProps> = (args) => {
  const [value, setValue] = React.useState<string>("");
  return <DatePicker {...args} setValue={setValue} value={value} />;
};

export const Default = Template.bind({});

export const Selected = Template.bind({});
Selected.args = {
  selected: new Date(2022, 5, 2),
};
Selected.play = ({ canvasElement }) => {
  const canvas = within(canvasElement);
  allowLayoutCalculations(() =>
    userEvent.click(canvas.getByRole("button"))
  );
};
