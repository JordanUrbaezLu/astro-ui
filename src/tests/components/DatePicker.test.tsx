import { render } from "@testing-library/react";
import { DatePicker } from "../../components/DatePicker/DatePicker";

describe("DatePicker", () => {
  test("Should render component correctly.", () => {
    const { container } = render(<DatePicker setValue={jest.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
