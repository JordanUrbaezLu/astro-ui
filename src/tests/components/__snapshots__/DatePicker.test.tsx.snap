// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker Should render component correctly. 1`] = `
<div>
  <div
    aria-label="textbox"
    class="container"
    role="textbox"
  >
    <div
      class="textFieldContainer"
    >
      <input
        aria-label="Select a Date"
        autocomplete="off"
        class="textField medium customInput"
        id="void_textfield"
        required=""
        spellcheck="false"
        type="text"
        value=""
      />
      <span
        class="label labelMedium"
      >
        Select a Date
      </span>
    </div>
    <button
      aria-label="Icon Button"
      class="iconButton medium primary"
    >
      <div
        class="container"
        role="presentation"
      />
      <svg
        fill="currentColor"
        height="22"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="22"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M112 880c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V460H112v420zm768-696H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v176h800V216c0-17.7-14.3-32-32-32z"
        />
      </svg>
    </button>
  </div>
</div>
`;
